File: calculator.c
#include <stdio.h>
#include <stdlib.h>
#include <ctype.h>
#include <string.h>
#include <math.h>
//global values
char stack[50];
int top = -1;

//get precedent
char pred(char exp) {
    if(exp == '+' || exp == '-') return 1;
    if(exp == '*' || exp == '/') return 2;
    if(exp == '^') return 3;
    return 0;
}

//push
void push(char val){
    stack[++top] = val;
}
//pop
char pop(){
    return stack[top--];
}
//get top value
char peek(){
    if (top == -1) return 0;
    return stack[top];
}

//infix to postfix
void postfix(char exp[], char postexp[]) {
    int j = 0;//pointer to exp
    int k = 0;//pointer to postexp
    while(exp[j] != '\0') {

        // keeping digits together
        if(isdigit(exp[j])) {

            while(isdigit(exp[j])){
                postexp[k++] = exp[j++];
            }
            postexp[k++] = ' '; //delimiter
        }
        
        //pushing (
        else if(exp[j] == '(') push(exp[j++]);

        //pop till (
        else if(exp[j] == ')'){
            while (peek() != '('){
                postexp[k++] = pop();
                postexp[k++] = ' '; 
            }
            pop(); //popping 
            j++;
        }
        
        //noise
        else if (exp[j] == ' ' || exp[j] == '\n' || exp[j] == '\t') {
            j++;
        }

        //operators
        else if (exp[j] == '+' || exp[j] == '-' || exp[j] == '*' || exp[j] == '/' || exp[j] == '^') {
            while (top != -1 && pred(peek()) >= pred(exp[j])) {
                postexp[k++] = pop();
                postexp[k++] = ' ';
            }
            push(exp[j++]);
        } 
    }

    //if values remaining 
    while(top != -1) {
        postexp[k++] = pop();
        postexp[k++] = ' ';
    }
    postexp[k] = '\0';


}
//evaluation of postfix
int evaluate(char postexp[]) {
    
    int i = 0;
    int itop = -1; //internal top
    int numstack[50];
    
    while(postexp[i] != '\0'){
        
        //removing whitespace
        if(postexp[i] == ' ') {
            i++;
            continue;
        }

        //storing digits in stack
        if(isdigit(postexp[i])) {
            int num = 0;
            while(isdigit(postexp[i])) {
                num = num * 10 + (postexp[i] - '0');
                i++;
            }
            numstack[++itop] = num;
        }

        //operations
        else {
            int val1 = numstack[itop--];
            int val2 = numstack[itop--];
        
             switch (postexp[i]) {
                case '+': numstack[++itop] = val1 + val2; break;
                case '-': numstack[++itop] = val1 - val2; break;
                case '*': numstack[++itop] = val1 * val2; break;
                case '/': numstack[++itop] = val1 / val2; break;
               // case '^': numstack[++itop] = pow(val1, val2); break;
            }
            i++;
        }
    }
    return numstack[itop];
}


int main() {
    char exp[50];
    char postexp[100];
    
    //Entering values
    printf("Enter string: ");
    fgets(exp, sizeof(exp), stdin);
    
    //cleaning up end
    exp[strcspn(exp, "\n")] = '\0';
    printf("Given expression: %s\n", exp);
    
    //validating postfix
    postfix(exp,postexp);
    printf("Postfix expression: %s\n", postexp);
    
    //final output
    int ans = evaluate(postexp);
    printf("Answer: %d\n", ans);
    
    return 0;
}

File: calculator.java
import java.util.*;
// import java.lang.*;

class calculator{
   
    // Getting precedent
    static int perd(char op) {
        if (op == '+' || op == '-') return 1;
        if (op == '*' || op == '/') return 2;
        if (op == '^') return 3;
        return 0;
    }

    //Infix to Postfix
    static String postfix(String exp) {
        StringBuilder postexp = new StringBuilder();
        Stack<Character> stack = new Stack<>();

        int i = 0; //index

        //Loop till lenght
        while (i < exp.length()) {

            //White spaces remove
            if (Character.isWhitespace(exp.charAt(i))) {
                i++;
                continue;
            }

            //Digits
            if (Character.isDigit(exp.charAt(i))) {
                while (i < exp.length() && Character.isDigit(exp.charAt(i))) {
                    postexp.append(exp.charAt(i));
                    i++;
                } 
                postexp.append(' ');
            }

            // ( 
            else if (exp.charAt(i) == '(') {
                stack.push(exp.charAt(i));
                i++;
            }

            // Popping till (
            else if (exp.charAt(i) == ')') {
                while(!stack.isEmpty() && stack.peek() != '(') {
                    postexp.append(stack.pop()).append(' ');
                }
                stack.pop();
                i++;
            }

            //oprands
            else {
                while (!stack.isEmpty() && perd(stack.peek()) >= perd(exp.charAt(i))){
                    postexp.append(stack.pop()).append(' ');
                }
                stack.push(exp.charAt(i));
                i++;
            }
        }

        //Emtying stack
        while (!stack.isEmpty()) {
           postexp.append(stack.pop()).append(' ');
        }

        return postexp.toString();
    }

    //Postfix Evaluation
    static int evaluation(String post) {
        
        Stack<Integer> stack = new Stack<>();
        int i = 0;

        while (i < post.length()) {

            if (post.charAt(i) == ' ') {
                i++;
                continue;
            }

            // number
            if (Character.isDigit(post.charAt(i))) {
                int num = 0;
                while (i < post.length() && Character.isDigit(post.charAt(i))) {
                    num = num * 10 + (post.charAt(i) - '0');
                    i++;
                }
                stack.push(num);
            }

            // operator
            else {
                int b = stack.pop();
                int a = stack.pop();

                switch (post.charAt(i)) {
                    case '+': stack.push(a + b); break;
                    case '-': stack.push(a - b); break;
                    case '*': stack.push(a * b); break;
                    case '/': stack.push(a / b); break;
                    case '^': stack.push((int)Math.pow(a, b)); break;
                }
                i++;
            }
        }

        return stack.peek();
    } 

    public static void main(String args[]) {
        Scanner sc = new Scanner(System.in);

        System.out.print("Enter Expresion: ");
        String exp = sc.nextLine();
        String postexp = postfix(exp);

        int result = evaluation(postexp);

        System.out.println("Entered Experation: " + exp);
        
        System.out.println("Postfix Expresion: " + postfix(exp));

        System.out.println("Answer = " + result);
        
        sc.close();
    }
}